caravel/model/full_text.py

```diff
@@ -0,0 +1,64 @@
+from google.appengine.ext import ndb
+from google.appengine.api import memcache
+import json
+
+FTS = "fts:"
+
+def tokenize(value):
+    """Parses the given string into words."""
+    return value.lower().split() + [""]
+
+
+class FullTextMixin(ndb.Model):
+    keywords = ndb.ComputedProperty(
+        lambda self: [""] + list(set(self._keywords())), repeated=True)
+
+    @classmethod
+    def matching(klass, query, offset=0, limit=None):
+        """Returns entities matching the given search query."""
+
+        # Check contents of cache.
+        words = tokenize(query)
+        results = memcache.get_multi(words, key_prefix=FTS)
+        matches = None
+        in_order = None
+        writeback = {}
+
+        # Fill anything missing from the initial fetch.
+        for word in words:
+            if word in results:
+                urlsafe = json.loads(results[word])
+                keys = [ndb.Key(urlsafe=val) for val in urlsafe]
+            else:
+                query = klass.query(klass.keywords == word)
+                keys = query.order(-klass.posted_at).fetch(keys_only=True)
+                writeback[word] = json.dumps([key.urlsafe() for key in keys])
+
+            matches = (matches & set(keys)) if matches else set(keys)
+
+        # Write back modified cache.
+        memcache.set_multi(writeback, time=3600, key_prefix=FTS)
+
+        # Elide potentially stale entries from the cache.
+        for candidate in ndb.get_multi([key for key in keys if key in matches]):
```

Comment:
We're doing something sorta tricky. (First, `keys` will never be null since words is non-empty.) Each `keys` is in order, so we iterate through one of them — but remove any that aren't in all matches (`if key in matches`). Hence we get a list of keys, in order, that contain each word.

I'll see if I can add a comment to that effect since it's certainly not explained well as-is.